packages/nimble-components/src/table/index.ts

```diff
@@ -66,15 +103,71 @@ export class Table<
 
         // Ignore any updates that occur prior to the TanStack table being initialized.
         if (this.tableInitialized) {
-            this.updateTableOptions({ data: this.data });
+            this.trySetData(this.data);
+        }
+    }
+
+    public checkValidity(): boolean {
+        return Object.values(this.validity).every(x => x === false);
+    }
+
+    private trySetData(newData: TData[]): void {
+        const areIdsValid = this.validateRecordIds();
+        if (areIdsValid) {
+            this.updateTableOptions({ data: newData });
+        } else {
+            this.updateTableOptions({ data: [] });
         }
     }
 
+    private validateRecordIds(): boolean {
+        // Start off by assuming everything is valid.
+        this.duplicateRowId = false;
+        this.missingRowId = false;
+        this.invalidRowId = false;
+
+        if (this.idFieldName == null) {
+            return true;
+        }
+
+        const ids = new Set<string>();
+        for (const record of this.data) {
+            if (
+                !Object.prototype.hasOwnProperty.call(record, this.idFieldName)
+            ) {
+                this.missingRowId = true;
+                return false;
```

Comment:
Is there value in not returning early in case the ids have multiple problems?

Reply:
The value is that we aren't doing extra work when we already know that the configuration is invalid, but I don't mind updating this to report all the problems. I was on the fence myself about this.

Reply:
I also don't feel too strongly, but I think I'd vote to continue validation beyond the first error. In the case we're considering, the table's already in a bad state so performance is not too important. And if we can save a developer from multiple rounds of debugging to discover all the errors then that's worthwhile.

Reply:
Updated & added a test